qiskit_nature/second_q/mappers/interleaved_qubit_mapper.py

```diff
@@ -0,0 +1,84 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A QubitMapper wrapper to transform from block-ordered to interleaved qubits."""
+
+from __future__ import annotations
+
+from qiskit.opflow import PauliSumOp
+from qiskit.quantum_info import SparsePauliOp
+
+from qiskit_nature.second_q.operators import SparseLabelOp
+
+from .qubit_mapper import QubitMapper
+
+
+class InterleavedQubitMapper(QubitMapper):
+    """A ``QubitMapper`` wrapper returned interleaved-ordered operators.
+
+    This class is intended to be used with fermionic systems. By convention, Qiskit Nature always
+    arranges the qubits corresponding to the alpha-spin and beta-spin components in a blocked
+    fashion. I.e. the first half of the qubit register corresponds to the alpha-spin components, and
+    the second half to the beta-spin one, like so:
+
+    .. code-block::
+
+       a1, a2, ..., aN, b1, b2, ..., bN
+
+    This class allows you to wrap any ``QubitMapper`` such that the produced qubit operators will
+    produce an interleaved order of qubits, instead. Taking the example from before, the outcome
+    will be the following:
+
+    .. code-block::
+
+       a1, b1, a2, b2, ..., aN, bN
+
+    Here is a very simple usage example:
+
+    .. code-block:: python
+
+       from qiskit_nature.second_q.mappers import JordanWignerMapper, InterleavedQubitMapper
+       from qiskit_nature.second_q.operators import FermionicOp
+
+       blocked_mapper = JordanWignerMapper()
+       interleaved_mapper = InterleavedQubitMapper(blocked_mapper)
+
+       ferm_op = FermionicOp({"+_0 -_1": 1}, num_spin_orbitals=4)
+
+       blocked_op = blocked_mapper.map(ferm_op)
+       # SparsePauliOp(['IIXY', 'IIYY', 'IIXX', 'IIYX'], coeffs=[-0.25j, 0.25, 0.25, 0.25j])
+
+       print(interleaved_mapper.map(ferm_op))
+       # SparsePauliOp(['IXIY', 'IYIY', 'IXIX', 'IYIX'], coeffs=[-0.25j, 0.25, 0.25, 0.25j])
+
+    """
+
+    def __init__(self, mapper: QubitMapper):
+        """
+        Args:
+            mapper: the actual ``QubitMapper`` mapping second-quantization operators to qubit
+                operators.
+        """
+        super().__init__(allows_two_qubit_reduction=mapper.allows_two_qubit_reduction)
+        self.mapper = mapper
+
+    def _map_single(self, second_q_op: SparseLabelOp) -> PauliSumOp:
+        blocked_op = self.mapper._map_single(second_q_op).primitive
```

Comment:
Note sure what happened to my comment - I know I wrote something like the following before!

Although this is intended for fermionic systems here its general enough that this order alteration could be done with any operator. Would would want do a size check (size of labels i,e number of qubits) to ensure the operator has an even size and if not raise an error. Otherwise I think the code will fail with some index error and maybe more obscure for a user to figure what went wrong.

Reply:
I think you commented on the commit itself [here](https://github.com/mrossinek/qiskit-nature/commit/7a31a4ee1ac3daa342b688a74c28628e80dca396#r98322174). Anyways, I suppose we can add a check like that. Will do this next week :+1: 

Should we also be adding more checks like whether or not the `mapper` provided to `__init__` is a `FermionicMapper`? That should also deal with operators other than `FermionicOp` being used here.

Reply:
Since this is just a changing order in the operator it could be done for any system right. Not sure why one would use it outside of a fermionic system and even there BKSF does interactions. So while we could restrict it, if you think thats desirable, things won't fail. If you use this mapper you might need to account for the fact if you have some custom circuits you built you used - U(VCC) would get taken care of automatically, as you, note via the mapper use.

Reply:
Indeed, this could be done for any kind, but the naming of this class specifically shows the intend of going from block-ordered fermionic spin orbitals to interleaved ordered ones. Thus, I think it is important to do this restriction because we do not _want_ to guarantee any behavior in the other cases.

To make this more clear, I updated the class to subclass `FermionicMapper` directly

Reply:
Do you want to check and deny its use with BKSF - while that might map an even number of orbitals to even number of qubits, since its mapping is done on interactions, not sites, that may not occur. One could check the qubit size resultant from the mapping, rather than the number or orbitals. That would mean an explicit check to BKSF was not needed - and reordering if the size was even would do no harm I think.

Reply:
Yeh I had not thought of that. I had hoped to do the check before the mapping in order to avoid the unnecessary mapping computation if it is bound to fail.

I suppose we can either:
- check the operator length on the qubit space
- expose `.mapper` via a property getter/setter and disallow BKSF in there

What do you think?

Reply:
A third option would be to check the self.mapper type at the same time as the existing orbital check there. You could change the mapper property to potentially detect things earlier, but since BKSF scales by interactions its uncommon for it to ever be used. I think we should note either way this interleaved mapper does not support BKSF in the docstring somewhere.

Reply:
The point is that this check is done _for every operator_ which seems like an annoying overhead to have to re-check the same mapper every time.